File: source/downloads/code/LeetCode/Python/Subsets2.py
class Solution(object):
    def subset_helper(self, nums, result, vector):
        result.append(vector[:])

        for i in range(len(nums)):
            # duplicated
            if i>0 and nums[i] == nums[i-1]:
                continue

            vector.append(nums[i])
            self.subset_helper(nums[i+1:],result,vector)
            vector.pop()

    def subsetsWithDup(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        if nums == None:
            return result
        self.subset_helper(sorted(nums), result,[])
        return result

File: source/downloads/code/LeetCode/Java/Subsets2.java
public class Solution {
    List<List<Integer>> ret = new ArrayList<List<Integer>> ();
    public List<List<Integer>> subsetsWithDup(int[] nums) {
        if (nums==null || nums.length==0)
            return ret;
        Arrays.sort(nums);
        // add [] into result
        result.add(new ArrayList<Integer>());
        int begin = 0;
        for(int i = 0; i < nums.length; i++){
                // find non-duplicate
                if (i==0 || nums[i] != nums[i-1])
                    begin = 0;
                int size = result.size();
                for (int j=begin; j<size; j++){
                    List<Integer> cur = new ArrayList<Integer>(result.get(j));
                    cur.add(nums[i]);
                    result(cur);
                }
                begin = size;
        }
        
        return ret;
    }
}

